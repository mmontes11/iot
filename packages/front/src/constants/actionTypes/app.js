export const TOGGLE_HAMBURGER_MENU = "TOGGLE_HAMBURGER_MENU";
export const TOGGLE_MAP_DIALOG = "TOGGLE_MAP_DIALOG";
export const SET_SHOW_ERROR = "SET_SHOW_ERROR";
