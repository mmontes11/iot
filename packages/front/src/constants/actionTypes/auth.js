export const IS_AUTH = "IS_AUTH";
export const SET_USERNAME = "SET_USERNAME";
export const SET_PASSWORD = "SET_PASSWORD";
export const LOGIN_REQUEST = "LOGIN_REQUEST";
export const LOGIN_REQUEST_SUCCESS = "LOGIN_REQUEST_SUCCESS";
export const LOGIN_REQUEST_ERROR = "LOGIN_REQUEST_ERROR";
