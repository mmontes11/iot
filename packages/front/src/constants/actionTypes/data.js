export const DATA_UPDATED = "DATA_UPDATED";
export const DATA_REQUEST = "DATA_REQUEST";
export const DATA_REQUEST_SUCCESS = "DATA_REQUEST_SUCCESS";
export const DATA_REQUEST_ERROR = "DATA_REQUEST_ERROR";
export const ADD_DATA_ITEM = "ADD_DATA_ITEM";
export const RESET_DATA = "RESET_DATA";
