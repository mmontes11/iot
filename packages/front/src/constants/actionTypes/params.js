export const PARAM_SELECT = "PARAM_SELECT";
export const PARAM_UPDATE = "PARAM_UPDATE";
export const PARAM_RESET = "PARAM_RESET";
export const PARAM_REQUEST = "PARAM_REQUEST";
export const PARAM_REQUEST_SUCCESS = "PARAM_REQUEST_SUCCESS";
export const PARAM_REQUEST_ERROR = "PARAM_REQUEST_ERROR";
export const PARAM_ITEMS_UPDATED = "PARAM_ITEMS_UPDATED";
export const PARAM_DISABLE = "PARAM_DISABLE";
