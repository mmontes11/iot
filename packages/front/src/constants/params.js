export const TYPE = "TYPE";
export const OBSERVATION = "OBSERVATION";
export const GROUPBY = "GROUPBY";
export const THING = "THING";
