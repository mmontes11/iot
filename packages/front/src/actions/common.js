import { RESET } from "constants/actionTypes/common";

export const reset = preserveError => dispatch => {
  dispatch({ type: RESET, preserveError });
};
