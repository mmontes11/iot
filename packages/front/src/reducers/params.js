import {
  PARAM_SELECT,
  PARAM_UPDATE,
  PARAM_RESET,
  PARAM_REQUEST,
  PARAM_REQUEST_SUCCESS,
  PARAM_REQUEST_ERROR,
  PARAM_ITEMS_UPDATED,
  PARAM_DISABLE,
} from "constants/actionTypes/params";
import { RESET } from "constants/actionTypes/common";
import deepEqual from "deep-equal";

export const initialState = {
  byId: {},
  allIds: [],
  isResetDisabled: true,
};

const paramInitialState = {
  items: [],
  isActive: false,
  isLoading: false,
  isDisabled: false,
  selectedItem: null,
};

const addParamIfNeeded = (state, paramId) => {
  if (state.allIds.includes(paramId)) {
    return state;
  }
  return {
    ...state,
    byId: {
      ...state.byId,
      [paramId]: paramInitialState,
    },
    allIds: [...state.allIds, paramId],
  };
};

const updateParam = (state, paramId, param) => ({
  ...state,
  byId: {
    ...state.byId,
    [paramId]: param,
  },
});

const allParamsInInitialState = state => Object.values(state.byId).every(param => deepEqual(param, paramInitialState));

export default (state = initialState, { type, param: paramId, selectedItem, items }) => {
  if (type === RESET) {
    return initialState;
  }
  if (!paramId) {
    return state;
  }
  let newState = { ...state };
  newState = addParamIfNeeded(newState, paramId);
  const param = newState.byId[paramId];
  switch (type) {
    case PARAM_SELECT:
      return updateParam(newState, paramId, { ...param, isActive: !param.isActive });
    case PARAM_UPDATE:
      return { ...updateParam(newState, paramId, { ...param, isActive: false, selectedItem }), isResetDisabled: false };
    case PARAM_RESET: {
      const newStateWithResetParam = updateParam(newState, paramId, paramInitialState);
      return { ...newStateWithResetParam, isResetDisabled: allParamsInInitialState(newStateWithResetParam) };
    }
    case PARAM_REQUEST:
      return updateParam(newState, paramId, { ...param, isLoading: true, isDisabled: true });
    case PARAM_REQUEST_SUCCESS:
      return updateParam(newState, paramId, { ...param, isLoading: false, isDisabled: false });
    case PARAM_REQUEST_ERROR:
      return updateParam(newState, paramId, { ...param, isLoading: false, isDisabled: false });
    case PARAM_ITEMS_UPDATED:
      return updateParam(newState, paramId, { ...param, items });
    case PARAM_DISABLE:
      return updateParam(newState, paramId, { ...param, isDisabled: true });
    default:
      return state;
  }
};

export const getParam = (state, paramName) => state.byId[paramName] || {};
