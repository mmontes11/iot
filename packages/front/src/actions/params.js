import {
  PARAM_SELECT,
  PARAM_UPDATE,
  PARAM_RESET,
  PARAM_REQUEST,
  PARAM_REQUEST_SUCCESS,
  PARAM_REQUEST_ERROR,
  PARAM_ITEMS_UPDATED,
  PARAM_DISABLE,
} from "constants/actionTypes/params";
import { EVENT_TYPE, MEASUREMENT_TYPE, OBSERVATION_TYPES } from "constants/observationTypes";
import { TYPE, OBSERVATION, GROUPBY, THING } from "constants/params";
import iotClient from "lib/iotClient";
import * as fromState from "reducers";
import { isEmpty } from "helpers/validation";

const requestObservations = (type, onStart, onSuccess, onError) => {
  if (type === EVENT_TYPE) {
    onStart();
    iotClient.eventService
      .getTypes()
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  } else if (type === MEASUREMENT_TYPE) {
    onStart();
    iotClient.measurementService
      .getTypes()
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  }
};

const requestObservationsDispatchingActions = (dispatch, type, cb) => {
  if (!type) {
    cb();
    return;
  }
  requestObservations(
    type,
    () => dispatch({ type: PARAM_REQUEST, param: OBSERVATION }),
    res => {
      dispatch({ type: PARAM_REQUEST_SUCCESS, param: OBSERVATION, statusCode: res.statusCode, error: null });
      dispatch({ type: PARAM_RESET, param: OBSERVATION });
      dispatch({ type: PARAM_ITEMS_UPDATED, param: OBSERVATION, items: res.body.types });
      cb();
    },
    error => {
      dispatch({ type: PARAM_REQUEST_ERROR, param: OBSERVATION, statusCode: error.statusCode, error });
      cb();
    },
  );
};

export const selectType = () => dispatch => {
  dispatch({ type: PARAM_ITEMS_UPDATED, param: TYPE, items: OBSERVATION_TYPES });
  dispatch({ type: PARAM_SELECT, param: TYPE });
};

export const updateType = type => dispatch => {
  dispatch({ type: PARAM_UPDATE, param: TYPE, selectedItem: type });
  requestObservationsDispatchingActions(dispatch, type, () => undefined);
};

export const selectObservation = () => (dispatch, getState) => {
  const state = getState();
  const observation = fromState.getParam(state, OBSERVATION);
  const dispatchSelectObservation = () => dispatch({ type: PARAM_SELECT, param: OBSERVATION });
  if (observation && !isEmpty(observation.items)) {
    dispatchSelectObservation();
  } else {
    const type = fromState.getParam(state, TYPE);
    if (type && type.selectedItem) {
      requestObservationsDispatchingActions(dispatch, type.selectedItem, dispatchSelectObservation);
    }
  }
};

export const updateObservation = observation => dispatch => {
  dispatch({ type: PARAM_UPDATE, param: OBSERVATION, selectedItem: observation });
};

export const selectGroupBy = () => (dispatch, getState) => {
  const state = getState();
  const groupBy = fromState.getParam(state, GROUPBY);
  if (groupBy && !isEmpty(groupBy.items)) {
    dispatch({ type: PARAM_SELECT, param: GROUPBY });
  } else {
    dispatch({ type: PARAM_REQUEST, param: GROUPBY });
    iotClient.timePeriodsService
      .getSupportedTimePeriods()
      .then(res => {
        dispatch({ type: PARAM_REQUEST_SUCCESS, param: GROUPBY, statusCode: res.statusCode, error: null });
        dispatch({ type: PARAM_ITEMS_UPDATED, param: GROUPBY, items: res.body.timePeriods });
        dispatch({ type: PARAM_SELECT, param: GROUPBY });
      })
      .catch(error => {
        dispatch({ type: PARAM_REQUEST_ERROR, param: GROUPBY, statusCode: error.statusCode, error });
      });
  }
};

export const updateGroupBy = groupBy => dispatch => {
  dispatch({ type: PARAM_UPDATE, param: GROUPBY, selectedItem: groupBy });
};

export const updateParams = params => dispatch => {
  const [firstParam, ...rest] = Object.keys(params);
  dispatch({ type: PARAM_UPDATE, param: firstParam, selectedItem: params[firstParam] });
  rest.forEach(key => {
    dispatch({ type: PARAM_DISABLE, param: key });
    dispatch({ type: PARAM_UPDATE, param: key, selectedItem: params[key] });
  });
};

export const selectThing = () => (dispatch, getState) => {
  const thing = fromState.getParam(getState(), THING);
  if (thing && !isEmpty(thing.items)) {
    dispatch({ type: PARAM_SELECT, param: THING });
  } else {
    dispatch({ type: PARAM_REQUEST, param: THING });
    iotClient.thingsService
      .getThings(true)
      .then(res => {
        dispatch({ type: PARAM_REQUEST_SUCCESS, param: THING, statusCode: res.statusCode, error: null });
        dispatch({ type: PARAM_ITEMS_UPDATED, param: THING, items: res.body.things });
        dispatch({ type: PARAM_SELECT, param: THING });
      })
      .catch(error => {
        dispatch({ type: PARAM_REQUEST_ERROR, param: THING, statusCode: error.statusCode, error });
      });
  }
};

export const updateThing = thingName => (dispatch, getState) => {
  dispatch({ type: PARAM_UPDATE, param: THING, selectedItem: thingName });
  const thingParam = fromState.getParam(getState(), THING);
  const thing = thingParam.items.find(item => item.name === thingName);
  dispatch({ type: PARAM_RESET, param: TYPE });
  dispatch({ type: PARAM_ITEMS_UPDATED, param: TYPE, items: thing.supportedObservationTypes.measurement });
};

export const selectMeasurementType = () => dispatch => {
  dispatch({ type: PARAM_SELECT, param: TYPE });
};

export const updateMeasurementType = type => dispatch => {
  dispatch({ type: PARAM_UPDATE, param: TYPE, selectedItem: type });
};
