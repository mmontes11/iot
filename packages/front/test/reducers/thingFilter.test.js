import deepFreeze from "deep-freeze";
import thingFilterReducer from "reducers/thingFilter";
import {
  THING_FILTER_SELECT,
  THING_FILTER_UPDATED,
  THING_FILTERS_REQUEST,
  THING_FILTERS_REQUEST_SUCCESS,
  THING_FILTERS_REQUEST_ERROR,
  THING_FILTERS_UPDATED,
} from "constants/actionTypes/thingFilter";
import { TYPE_SELECT } from "constants/actionTypes/params";
import { RESET } from "constants/actionTypes/common";

import { initialState } from "../constants/index";

const thingFilterInitialState = initialState.filters.thingFilter;
deepFreeze(thingFilterInitialState);

describe("reducers/thingFilter", () => {
  it("has a default state", () => {
    expect(thingFilterReducer(undefined, { type: "@init" })).toMatchSnapshot();
  });
  it("reduces nothing", () => {
    expect(thingFilterReducer(thingFilterInitialState, { type: "WHATEVER" })).toMatchSnapshot();
  });
  it("reduces TYPE_SELECT", () => {
    expect(thingFilterReducer(thingFilterInitialState, { type: TYPE_SELECT })).toMatchSnapshot();
  });
  it("reduces THING_FILTER_SELECT", () => {
    expect(thingFilterReducer(thingFilterInitialState, { type: THING_FILTER_SELECT })).toMatchSnapshot();
  });
  it("reduces THING_FILTER_UPDATED", () => {
    expect(thingFilterReducer(thingFilterInitialState, { type: THING_FILTER_UPDATED })).toMatchSnapshot();
  });
  it("reduces THING_FILTERS_REQUEST", () => {
    expect(thingFilterReducer(thingFilterInitialState, { type: THING_FILTERS_REQUEST })).toMatchSnapshot();
  });
  it("reduces THING_FILTERS_REQUEST_SUCCESS", () => {
    expect(thingFilterReducer(thingFilterInitialState, { type: THING_FILTERS_REQUEST_SUCCESS })).toMatchSnapshot();
  });
  it("reduces THING_FILTERS_REQUEST_ERROR", () => {
    expect(thingFilterReducer(thingFilterInitialState, { type: THING_FILTERS_REQUEST_ERROR })).toMatchSnapshot();
  });
  it("reduces THING_FILTERS_UPDATED", () => {
    expect(thingFilterReducer(thingFilterInitialState, { type: THING_FILTERS_UPDATED })).toMatchSnapshot();
  });
  it("reduces RESET", () => {
    expect(thingFilterReducer(thingFilterInitialState, { type: RESET })).toMatchSnapshot();
  });
});
