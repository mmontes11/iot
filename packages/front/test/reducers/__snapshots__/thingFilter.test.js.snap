// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/thingFilter has a default state 1`] = `
Object {
  "isActive": false,
  "isDisabled": false,
  "isLoading": false,
  "items": Array [],
  "selectedItem": null,
}
`;

exports[`reducers/thingFilter reduces RESET 1`] = `
Object {
  "isActive": false,
  "isDisabled": false,
  "isLoading": false,
  "items": Array [],
  "selectedItem": null,
}
`;

exports[`reducers/thingFilter reduces THING_FILTER_SELECT 1`] = `
Object {
  "isActive": true,
  "isDisabled": false,
  "isLoading": false,
  "items": Array [],
  "selectedItem": null,
}
`;

exports[`reducers/thingFilter reduces THING_FILTER_UPDATED 1`] = `
Object {
  "isActive": false,
  "isDisabled": false,
  "isLoading": false,
  "items": Array [],
  "selectedItem": undefined,
}
`;

exports[`reducers/thingFilter reduces THING_FILTERS_REQUEST 1`] = `
Object {
  "isActive": false,
  "isDisabled": true,
  "isLoading": true,
  "items": Array [],
  "selectedItem": null,
}
`;

exports[`reducers/thingFilter reduces THING_FILTERS_REQUEST_ERROR 1`] = `
Object {
  "isActive": false,
  "isDisabled": false,
  "isLoading": false,
  "items": Array [],
  "selectedItem": null,
}
`;

exports[`reducers/thingFilter reduces THING_FILTERS_REQUEST_SUCCESS 1`] = `
Object {
  "isActive": false,
  "isDisabled": false,
  "isLoading": false,
  "items": Array [],
  "selectedItem": null,
}
`;

exports[`reducers/thingFilter reduces THING_FILTERS_UPDATED 1`] = `
Object {
  "isActive": true,
  "isDisabled": false,
  "isLoading": false,
  "items": undefined,
  "selectedItem": null,
}
`;

exports[`reducers/thingFilter reduces TYPE_SELECT 1`] = `
Object {
  "isActive": false,
  "isDisabled": false,
  "isLoading": false,
  "items": Array [],
  "selectedItem": null,
}
`;

exports[`reducers/thingFilter reduces nothing 1`] = `
Object {
  "isActive": false,
  "isDisabled": false,
  "isLoading": false,
  "items": Array [],
  "selectedItem": null,
}
`;
